debezium/dbz#1960: simple LOB partial rollback implementation (#7472)
debezium/debezium: documentation/modules/ROOT/pages/connectors/oracle.adoc

For the connector to use the `heartbeat.action.query` with tables that are not owned by the xref:creating-users-for-the-connector[connector user account], you must grant the connector user permission to run the necessary `INSERT` or `UPDATE` queries on those tables.
====

// Type: concept
// ModuleID: how-debezium-captures-large-object-data
// Title: How {prodname} captures large object data
[[capturing-large-object-data]]
=== Capturing large object data

Oracle stores several data types by using internal large object (LOB) mechanisms in the redo logs.

> To capture changes for any of these types, you must set the xref:oracle-property-lob-enabled[`lob.enabled`] connector configuration property to `true`.

Reviewer: ```suggestion
To permit {prodname} to capture changes to LOB data, you must set the xref:oracle-property-lob-enabled[`lob.enabled`] connector configuration property to `true`.
```

Reviewer 2: @roldanbob I applied all the edits, but the one thing that stood out to me from the edits was this sentence and what follows. It reads:

> To permit {prodname} to capture changes to LOB data, you must set the xref:oracle-property-lob-enabled[`lob.enabled`] connector configuration property to `true`.
> 
> Set `lob.enabled` to `true` if you want the connector to capture the following data types:

Is it me, or does that seem repetitive to you?

I've committed the changes as suggested, so you can give it another pass, but just thought I'd check. If it seems good to you, I'm perfectly fine letting it go as-is. You're definitely the better expert than I.

> The following data types require `lob.enabled` to be set to `true`:

Reviewer: ```suggestion
Set `lob.enabled` to `true` if you want the connector to capture the following data types:
```

* `CLOB` - Character large objects
* `NCLOB` - National character set large objects
* `BLOB` - Binary large objects
* `XMLTYPE` - XML data stored as a LOB

> * `JSON` - JSON data (Oracle 21c and later)

Author: Here is how LogMiner sees
```sql
-- CREATE TABLE TESTJSON(id NUMBER(9) NOT NULL PRIMARY KEY, data JSON);
BEGIN
	SAVEPOINT s1;
    UPDATE TESTJSON SET data = '{"id":1,"v":2}' WHERE id = 1;
    ROLLBACK TO SAVEPOINT s1;
END;
```
on Oracle 26ai Free Release 23.26.2.0.0:
|SEQUENCE#|OPERATION  |ROLLBACK|ROW_ID            |RS_ID                   |SQL_REDO                   |
|---------|-----------|--------|------------------|------------------------|---------------------------|
|        1|START      |       0|AAASA6AAAAAAAAAAAA| 0x000002.0000c730.0010 |set transaction read write;|
|        1|INTERNAL   |       0|AAASA6AAAAAAAMOAAA| 0x000002.0000c730.0010 |                           |
|        1|UNSUPPORTED|       0|AAAAAAAAAAAAAAAAAA| 0x000002.0000c731.0064 |Unsupported                |
|        1|UNSUPPORTED|       0|AAASA6AAAAAAAMOAAA| 0x000002.0000c732.0030 |Unsupported                |
|        1|UNSUPPORTED|       1|AAASA6AAAAAAAMOAAA| 0x000002.0000c732.01d4 |Unsupported                |
|        1|INTERNAL   |       1|AAASA6AAAAAAAMOAAA| 0x000002.0000c733.0100 |                           |
|        1|COMMIT     |       0|AAAAAAAAAAAAAAAAAA| 0x000002.0000c733.01d4 |commit;                    |

I've not tested rollbacks with JSON at all.

Author: Interesting, it seems like all operations become UNSUPPORTED when table has a JSON column... Should I worry about it? :)

Reviewer 2: Oracle emits JSON like a typical `CLOB`, `NCLOB`, or `BLOB` data type, but with a slightly different preamble in the generated SQL. So for all intents and purposes, its synonmous with those data types. 

I only raise it here because we have a JSON section, and you see UNSUPPORTED above because you need to trigger GoldenGate replication with Oracle 21 and LogMiner to see the data. If you don't enable GG replication, LogMiner won't natively see JSON data and will report UNSUPPORTED; worse, XStream isn't allowed to emit JSON regardless. 

See https://debezium.io/documentation/reference/3.7/connectors/oracle.html#oracle-json-types

> * Extended strings - `VARCHAR2`, `NVARCHAR2`, and `RAW` columns that exceed the legacy maximum lengths when the database parameter `max_string_size` is set to `EXTENDED` (see xref:support-for-extended-max-string-size[Extended max string size])

Reviewer: ```suggestion
`CLOB`:: Character large objects.
`NCLOB`:: National character set large objects.
`BLOB`:: Binary large objects.
`XMLTYPE`:: XML data stored as a LOB.
`JSON`:: JSON data (Oracle 21c and later).
Extended strings:: `VARCHAR2`, `NVARCHAR2`, and `RAW` columns that exceed the legacy maximum lengths when the database parameter `max_string_size` is set to `EXTENDED`.
For more information about the connector supports extended strings, see xref:support-for-extended-max-string-size[Extended max string size].
```

When `lob.enabled` is `false` (the default), snapshot events include the current LOB column values, but streamed change events represent LOB columns with the xref:oracle-property-unavailable-value-placeholder[`unavailable.value.placeholder`] constant.

> No LOB-specific redo log entries are mined, which reduces the volume of data that LogMiner processes.

Reviewer: ```suggestion
Disabling mining of LOB values reduces the volume of data that LogMiner processes by excluding LOB-specific redo log entries.
```

When `lob.enabled` is `true`, the connector mines additional redo log operation types that Oracle uses to represent LOB writes.
Oracle only supplies LOB column values when the column value is explicitly set or changed in a SQL statement.

> If a LOB column is not modified by a given DML statement, the change event contains the unavailable value placeholder rather than the column's current value.

Reviewer: ```suggestion
If a DML statement does not modify an LOB column, the change event contains the configured unavailable value placeholder rather than the column's current value.
```

==== INTERNAL events and ROLLBACK TO SAVEPOINT handling

> Oracle uses `INTERNAL` events (operation code 0) in the redo logs to associate `ROW_ID` values with out-of-line LOB column operations.

Reviewer: ```suggestion
Oracle writes `INTERNAL` events  (operation code 0) to the redo log to associate `ROW_ID` values with out-of-line LOB operations.
Configure `log.mining.include.internal.events` to specify whether the connector captures these events to enable accurate handling of `ROLLBACK TO SAVEPOINT` in transactions that modify LOB columns.
```

When a table stores LOB data out-of-line, these events provide the connector with the row identifier that it needs to correctly correlate LOB changes with their parent DML operation.

This correlation becomes critical when a transaction uses `ROLLBACK TO SAVEPOINT` to partially undo changes.

> Without `ROW_ID` information from `INTERNAL` events, the connector cannot always determine which specific row a rolled-back LOB operation belongs to.

Reviewer: ```suggestion
In the absence of the `ROW_ID` information that an `INTERNAL` event provides, the connector cannot reliably determine which row a rolled-back LOB operation affects.
```

> In these scenarios, the connector may emit change events with stale LOB field values.

Reviewer: ```suggestion
In such scenarios, the connector might emit change events that contain stale LOB field values.
```

> To address this, the connector provides the xref:oracle-property-log-mining-include-internal-events[`log.mining.include.internal.events`] configuration property.

Reviewer: ```suggestion
To prevent {prodname} from emitting change events that contain stale LOB values, you can set the xref:oracle-property-log-mining-include-internal-events[`log.mining.include.internal.events`] configuration property.
```

* When set to `false` (the default), `INTERNAL` events are excluded from LogMiner queries.
This keeps the volume of mined redo log data lower, but partial rollbacks in transactions that modify out-of-line LOB columns may produce change events with stale values for the affected LOB fields.
* When set to `true`, the connector includes `INTERNAL` events in LogMiner queries.

> This enables the connector to accurately handle `ROLLBACK TO SAVEPOINT` for tables with out-of-line LOB columns, at the cost of increased redo log query volume.

Reviewer: ```suggestion
The `log.mining.include.internal.events` property accepts the following values:

`false`:: (default) The connector excludes `INTERNAL` events from LogMiner queries.
This setting reduces the volume of mined redo log data.
However, after a partial rollback of a transaction that modifies out-of-line LOB columns, the connector might emit change events that contain stale values for affected LOB fields.
 
`true`:: The connector includes `INTERNAL` events in LogMiner queries.
This setting increases the volume of mined redo log data, but enables the connector to reliably handle `ROLLBACK TO SAVEPOINT` operations for tables with out-of-line LOB columns.
```
In general, I'm trying to pare down the volume of hybrid concept/task topics in the docjs. We could make this a separate reference topic, but in the current concept topic context, let's stick to the "what" framing vs. the "how" framing.

For the connector to use the `heartbeat.action.query` with tables that are not owned by the xref:creating-users-for-the-connector[connector user account], you must grant the connector user permission to run the necessary `INSERT` or `UPDATE` queries on those tables.
====

// Type: concept
// ModuleID: how-debezium-captures-large-object-data
// Title: How {prodname} captures large object data
[[capturing-large-object-data]]
=== Capturing large object data

Oracle stores several data types by using internal large object (LOB) mechanisms in the redo logs.
To permit {prodname} to capture changes to LOB data, you must set the xref:oracle-property-lob-enabled[`lob.enabled`] connector configuration property to `true`.

> Set `lob.enabled` to `true` if you want the connector to capture the following data types:

Reviewer: ```suggestion
To enable {prodname} to capture changes to LOB data, you must set the xref:oracle-property-lob-enabled[`lob.enabled`] connector configuration property to `true`.

With this property enabled, the connector can capture the following data types:
```

`CLOB`:: Character large objects.
`NCLOB`:: National character set large objects.
`BLOB`:: Binary large objects.
`XMLTYPE`:: XML data stored as a LOB.
`JSON`:: JSON data (Oracle 21c and later).
Extended strings:: `VARCHAR2`, `NVARCHAR2`, and `RAW` columns that exceed the legacy maximum lengths when the database parameter `max_string_size` is set to `EXTENDED`.

> For more information about the connector supports extended strings, see xref:support-for-extended-max-string-size[Extended max string size].

Reviewer: ```suggestion
For more information about how the connector supports extended strings, see xref:support-for-extended-max-string-size[Extended max string size].
```

> When `lob.enabled` is `false` (the default), snapshot events include the current LOB column values, but streamed change events represent LOB columns with the xref:oracle-property-unavailable-value-placeholder[`unavailable.value.placeholder`] constant.

Reviewer: ```suggestion
When `lob.enabled` is set to `false` (the default), snapshot events include the current LOB column values, but streamed change events represent LOB columns with the xref:oracle-property-unavailable-value-placeholder[`unavailable.value.placeholder`] constant.
```

Disabling mining of LOB values reduces the volume of data that LogMiner processes by excluding LOB-specific redo log entries.

When `lob.enabled` is `true`, the connector mines additional redo log operation types that Oracle uses to represent LOB writes.
Oracle only supplies LOB column values when the column value is explicitly set or changed in a SQL statement.
If a DML statement does not modify an LOB column, the change event contains the configured unavailable value placeholder rather than the column's current value.

> ==== INTERNAL events and ROLLBACK TO SAVEPOINT handling

Reviewer: ```suggestion
// Type: concept
// TItle: How Oracle processes INTERNAL events and savepoint rollbacks
// ModuleID: how-oracle-processes-internal-events-and-savepoint-rollbacks
[id="oracle-internal-events-and-rollback-to-savepoint-handling"]
==== INTERNAL events and ROLLBACK TO SAVEPOINT handling
```
Added metadata for downstream processing

Author: There is a typo in "// T**I**tle: How..."

> Oracle writes `INTERNAL` events  (operation code 0) to the redo log to associate `ROW_ID` values with out-of-line LOB operations.

Reviewer: ```suggestion
Oracle records `INTERNAL` events in the redo log to track row associations (`ROW_ID` values) for out-of-line LOB operations. 
Capturing these events helps the connector avoid emitting stale LOB values after savepoint rollbacks.
```
Adds a short description to summarize the "what" and "why" of the topic.

Author: Personally, I prefer the original phrasing: "...to associate `ROW_ID` values with out-of-line LOB operations."

For the connector to use the `heartbeat.action.query` with tables that are not owned by the xref:creating-users-for-the-connector[connector user account], you must grant the connector user permission to run the necessary `INSERT` or `UPDATE` queries on those tables.
====

// Type: concept
// ModuleID: how-debezium-captures-large-object-data
// Title: How {prodname} captures large object data
[[capturing-large-object-data]]
=== Capturing large object data

> Oracle stores several data types by using internal large object (LOB) mechanisms in the redo logs.
> To enable {prodname} to capture changes to LOB data, you must set the xref:oracle-property-lob-enabled[`lob.enabled`] connector configuration property to `true`.

Reviewer: ```suggestion
Capturing Oracle LOB data during streaming requires special handling. 
Configure the connector to capture LOB values accurately while balancing mining overhead and performance.

To accurately capture LOB column values during streaming, set the `lob.enabled` connector configuration property to `true`.
```

With this property enabled, the connector can capture the following data types:

`CLOB`:: Character large objects.
`NCLOB`:: National character set large objects.
`BLOB`:: Binary large objects.
`XMLTYPE`:: XML data stored as a LOB.
`JSON`:: JSON data (Oracle 21c and later).
Extended strings:: `VARCHAR2`, `NVARCHAR2`, and `RAW` columns that exceed the legacy maximum lengths when the database parameter `max_string_size` is set to `EXTENDED`.
For more information about how the connector supports extended strings, see xref:support-for-extended-max-string-size[Extended max string size].

When `lob.enabled` is set to `false` (the default), snapshot events include the current LOB column values, but streamed change events represent LOB columns with the xref:oracle-property-unavailable-value-placeholder[`unavailable.value.placeholder`] constant.
Disabling mining of LOB values reduces the volume of data that LogMiner processes by excluding LOB-specific redo log entries.

When `lob.enabled` is `true`, the connector mines additional redo log operation types that Oracle uses to represent LOB writes.

> Oracle only supplies LOB column values when the column value is explicitly set or changed in a SQL statement.

Reviewer: ```suggestion
During streaming, Oracle includes LOB column values in redo data only when a SQL DML statement explicitly sets or changes an LOB column value.
```
Qualify statement to specify that it applies to the streaming phase and restate the sentence in active voice.

If a DML statement does not modify an LOB column, the change event contains the configured unavailable value placeholder rather than the column's current value.

// Type: concept
// Title: How Oracle processes INTERNAL events and savepoint rollbacks
// ModuleID: how-oracle-processes-internal-events-and-savepoint-rollbacks
[id="oracle-internal-events-and-rollback-to-savepoint-handling"]
==== INTERNAL events and ROLLBACK TO SAVEPOINT handling

Oracle records `INTERNAL` events in the redo log to associate `ROW_ID` values with out-of-line LOB operations.
Capturing these events helps the connector avoid emitting stale LOB values after savepoint rollbacks.

> Configure `log.mining.include.internal.events` to specify whether the connector captures these events to enable accurate handling of `ROLLBACK TO SAVEPOINT` in transactions that modify LOB columns.
> When a table stores LOB data out-of-line, these events provide the connector with the row identifier that it needs to correctly correlate LOB changes with their parent DML operation.

Reviewer: ```suggestion
Set the `log.mining.include.internal.events` property to specify whether the connector captures `INTERNAL` events. 
The connector requires these events to accurately process `ROLLBACK TO SAVEPOINT` operations in transactions that modify LOB columns.

When a table stores LOB data out-of-line, these events provide the connector with the row identifier that it needs to correctly correlate LOB changes with their parent DML operation.
```

This correlation becomes critical when a transaction uses `ROLLBACK TO SAVEPOINT` to partially undo changes.
In the absence of the `ROW_ID` information that an `INTERNAL` event provides, the connector cannot reliably determine which row a rolled-back LOB operation affects.
In such scenarios, the connector might emit change events that contain stale LOB field values.

> To prevent {prodname} from emitting change events that contain stale LOB values, you can set the xref:oracle-property-log-mining-include-internal-events[`log.mining.include.internal.events`] configuration property.

Reviewer: ```suggestion
The `log.mining.include.internal.events` property controls whether the connector includes Oracle `INTERNAL` events in LogMiner queries.

The `log.mining.include.internal.events` property accepts the following values:
```

> * When set to `false` (the default), `INTERNAL` events are excluded from LogMiner queries.
> This keeps the volume of mined redo log data lower, but partial rollbacks in transactions that modify out-of-line LOB columns may produce change events with stale values for the affected LOB fields.
> * When set to `true`, the connector includes `INTERNAL` events in LogMiner queries.
> The `log.mining.include.internal.events` property accepts the following values:
> `false`:: (default) The connector excludes `INTERNAL` events from LogMiner queries.
> This setting reduces the volume of mined redo log data.
> However, after a partial rollback of a transaction that modifies out-of-line LOB columns, the connector might emit change events that contain stale values for affected LOB fields.
> 
> `true`:: The connector includes `INTERNAL` events in LogMiner queries.
> This setting increases the volume of mined redo log data, but enables the connector to reliably handle `ROLLBACK TO SAVEPOINT` operations for tables with out-of-line LOB columns.

Reviewer: ```suggestion
`false`:: (Default) The connector excludes `INTERNAL` events from LogMiner queries.
This setting reduces the volume of mined redo log data.
However, after a partial rollback of a transaction that modifies out-of-line LOB columns, the connector might emit change events that contain stale values for affected LOB fields.

`true`:: The connector includes `INTERNAL` events in LogMiner queries.
This setting increases the volume of mined redo log data, but enables the connector to accurately process `ROLLBACK TO SAVEPOINT` operations for tables with out-of-line LOB columns.
```

[NOTE]
====
The `log.mining.include.internal.events` property has no effect unless `lob.enabled` is also set to `true`.
====

[TIP]
====
If your application does not use `ROLLBACK TO SAVEPOINT` in transactions that modify tables with LOB columns, you can safely leave `log.mining.include.internal.events` set to `false`.
Enable it only when you need accurate partial rollback handling for tables that store LOB data out-of-line.
====

> As an alternative, you can set `lob.enabled` to `false` and use the xref:post-processors/reselect-columns.adoc[Reselect columns post processor] to re-query LOB column values directly from the database.
> This approach avoids the overhead of mining both LOB-specific redo log entries and `INTERNAL` events entirely, while still ensuring that LOB fields contain correct values, since the post processor fetches the current column value from the source table.

Reviewer: ```suggestion
As an alternative to enabling LOB mining and capturing `INTERNAL` events, you can set `lob.enabled` to `false` and configure the xref:post-processors/reselect-columns.adoc[Reselect columns post processor] to re-query LOB column values directly from the database.

This approach avoids the overhead of mining LOB-specific redo log entries and `INTERNAL` events while still ensuring that change events contain correct LOB values, because the post processor retrieves the values directly from the source table.
```

This configuration is ignored when using the XStream or OpenLogReplicator adapters.
endif::community[]

|[[oracle-property-log-mining-include-internal-events]]<<oracle-property-log-mining-include-internal-events, `+log.mining.include.internal.events+`>>
|`false`
|Specifies whether the connector supports mining `INTERNAL` events.
When set to `false`, the default, `INTERNAL` events are not captured.
When `lob.enabled` is `true` and all LOB columns of an operation are stored out-of-line, `INTERNAL` events provide the `ROW_ID` values necessary for accurate `ROLLBACK TO SAVEPOINT` handling.

> Enabling this option might significantly increase the volume of events returned by LogMiner.

Reviewer: ```suggestion
|Specifies whether the connector captures `INTERNAL` events.

Set one of the following values:

`false`:: (default) The connector does not capture `INTERNAL` events.
Set this option to reduce the volume of redo log data that the connector processes.
After a partial rollback of a transaction that modifies externally stored LOB data, the change events that the connector emits might contain stale values for the affected LOB fields.
 
`true`:: The connector captures `INTERNAL` events.
Set this option to enable the connector to accurately handle `ROLLBACK TO SAVEPOINT` operations for tables with out-of-line LOB columns.
This option can result in an increased volume of redo log data.

For more information, see xref:capturing-large-object-data[Capturing large object data].
```
